### ISSUE TITLE
Recursive Snapshots

### ISSUE THREAD

**Issue by necouchman** [zfsonlinux/zfs#5819]
So, I'm using snapshots on ZFS on Linux, in a recursive fashion, and running into an issue that I believe is a bug, but may just be the way ZFS works.  When I have a parent/child data set, and create recursive snapshots, the data inside the snapshots of the child data sets does not appear in the parent snapshot.  So, for example, if I have:
/fstest
/fstest/iso
/fstest/home

If I create a recursive snapshot, snap1, I have the following snapshots:
/fstest@snap1
/fstest/iso@snap1
/fstest/home@snap1

If I access the hidden snapshot directory at the parent level, I would expect to be able to see the data from the matching child snapshots.  Instead, when I go to /fstest/.zfs/snapshot/snap1/iso, I see an empty folder - just the place where the child filesystem would be mounted.  I'm not sure if this is the way it is supposed to work - if I have to go to /fstest/iso/.zfs/snapshot/snap1 to see that snapshot data, or if it should be present in the parent .zfs/snapshot/snap1 data set?

I'm using ZFS 0.6.5.9 on CentOS 7.3.

**Comment by evan-king** [comment]
I don't believe that's a bug.  The recursive action is simply repeating/propagating the action of taking a snapshot into all datasets with a child relationship inferred from mount path.  Other than association by mount path, the datasets have no relationship or shared data with each other, and neither do the snapshots.

While it would be technically feasible to "project" the content of same-named snapshots in child datasets into the view of a snapshot for a parent dataset, such behavior would really only serve to confuse matters and complicate some interactions with that distinct snapshot/dataset (such as trying to selectively restore directories from it).  And we haven't even yet considered the scenario where a child dataset masks files stored under the same path but actually in the parent dataset.

In short, I believe you should adjust your expectations - look for your data directly in the snapshots of the dataset that actually contains it.

**Comment by kpande** [comment]
please use [the mailing list](http://list.zfsonlinux.org) or IRC (#zfsonlinux on freenode) for support requests, this not a bug.
